### ISSUE TITLE
Warning in Unmounted Component

### ISSUE THREAD

**Issue by r1ckyrockz** [CharlesStover/reactn#76]
Hi there,

after i update `reactn` from version `1.0.0` to `2.0.1` i get on every page change this warning:

```
index.js:xxx Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
```

This didn't happen before.

Here is an example of my implementation:

package.json:
```
"react": "^16.8.6",
"react-router": "^5.0.0",
"react-router-dom": "^5.0.0",
"redux": "^4.0.1",
"reactn": "^2.0.1",
"typescript": "3.4.5",
// also the types for each package
```

```
// index.tsx
import React from 'reactn';
import ReactDOM from 'react-dom';
import {BrowserRouter} from 'react-router-dom';

// typings made before
setGlobal({payments: [], invoices: []})
addReducer('getPayments', (global: State, dispatch: Dispatch) => 
   axios.get('url/payments'...).then((data) => ({payments: data})
);

addReducer('getInvoices', (global: State, dispatch: Dispatch) => 
   axios.get('url/invoices'...).then((data) => ({invoices: data})
);

ReactDOM.render(
        <BrowserRouter>
                <App />
            </BrowserRouter>,
        document.getElementById('root')
    );
```

```
// App.tsx
import React from 'reactn';

const Payments = props => {
  const [payments] = useGlobal('payments');
  return <div>....</div>;
}

const Invoices = props => {
  const [invoices] = useGlobal('invoices');
  return <div>....</div>;
}

class App extends React.Component {
  render() {
    return (
      <div>
         <Payments />
         <InvoiceList />
      </div>
    );
  }
}
```
Not sure if this is related to https://github.com/CharlesStover/reactn/issues/28

Thanks for your help

**Comment by CharlesStover** [comment]
Do you have a minimal reproduction? You say the error occurs on page change, but your provided code only has one page.

**Comment by r1ckyrockz** [comment]
My fault. Here is an example with routing:

```
// App.tsx
class App extends React.Component {
    render() {
        return (
            <Fragment>
                <Link to="/payments">Payment</Link>
                <Link to="/invoices">Payment</Link>
                <Switch>
                    <Route path="/payments" component={PaymentList} />
                    <Route path="/invoices" component={InvoiceList} />
                </Switch>
            </Fragment>
        );
    }
}

```

```
// PaymentList Component
const Payments = props => {
  const [payments] = useGlobal('payments');
  return <div>....</div>;
}
```

```
// InvoiceList Component
const Invoices = props => {
  const [invoices] = useGlobal('invoices');
  return <div>....</div>;
}
```

**Comment by CharlesStover** [comment]
I am running your minimal reproduction, but not encountering this error. It must be code that you are leaving out that causes it.

```JavaScript
// App.tsx
import React, { Fragment, setGlobal, useGlobal } from 'reactn';
import { Link, Route, Switch } from 'react-router-dom';

interface G {
  invoices: string;
  payments: string;
}

setGlobal<G>({
  invoices: 'yes',
  payments: 'no',
});

// PaymentList Component
const Payments = () => {
  const [payments] = useGlobal<G>('payments');
  return <div>{payments}</div>;
}

// InvoiceList Component
const Invoices = () => {
  const [invoices] = useGlobal<G>('invoices');
  return <div>{invoices}</div>;
}

// App.tsx
export default class App extends React.Component {
  render() {
    return (
      <Fragment>
        <Link to="/payments">Payment</Link>
        <Link to="/invoices">Invoices</Link>
        <Switch>
          <Route path="/payments" component={Payments} />
          <Route path="/invoices" component={Invoices} />
        </Switch>
      </Fragment>
    );
  }
}
```

**Comment by bugs181** [comment]
Strange, I'm seeing this error too. Unfortunately, I don't have a minimal reproducible example. I'm using the Material UI framework and using this bit of code to switch between pages:

```
const page = ({ match }) => {
  const Component = pages[match.params.pageName]

  if (!Component)
    return null

  return <Component />
}
```

On the component pages themselves, I'm not doing anything except creating some DOM elements and using `this.global` state with them. The error happens when switching between pages and is difficult to reproduce as it doesn't always happen. Mostly happens when state is being changed from a Websocket stream while switching pages.

**Comment by r1ckyrockz** [comment]
May be I found the problem 🎉

The state will be updated over reducer (axios call). I call it on `componentWillMount()`, because i want always the latest data.

If the state is equal as before updating, i get the described error. So i need to check, if the state is equal (i do it with `lodash.isequal`).

Example:
```
import isEqual from 'lodash.isequal';

....
addReducer('getInvoices', (global: State, dispatch: Dispatch) => 
   axios.get('url/invoices'...)
     .then((data) => {
        if (isEqual(data, global.invoices)){
          return null;
        }
        return ({invoices: data})
     })
);
...
```

Not the best choice, but it works :(

**Comment by CharlesStover** [comment]
For the sake of reproducibility, if you replace `axios.get('url/invoices',...)` with something like the following, does the error still happen reliably and does your change still fix it reliably?

```JavaScript
new Promise((resolve) => {
  setTimeout(() => {
    resolve({
      invoices: Math.random(),
    });
  }, 200);
})
  .then((data) => {
    // ...
  });
```

What does it look like when you call `getInvoices()`? Does it use async/await? Does it have a `.then()`? Is it on mount, or in a `useEffect`?

It is hard for me to reproduce your scenario without hitting your API or having the shape of your `invoices` response, so if we can fill it with mock data, the solution should be easier to find.

I just need a way to reproduce this locally.

**Comment by umbertoghio** [comment]
Hello, 

today I tested the upgrade from react 0.2.0 to 2.0.2 in an Expo React native application, and I am getting several of this warnings when changing page (using react-navigation).

I tried to dig in a bit and this are my findings:

Error: 

`Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.%s, 
    in Tickets (at SceneView.js:9)
    in SceneView (at createTabNavigator.js:39)
    in RCTView (at View.js:44)
    in RCTScrollContentView (at ScrollView.js:852)
    in RCTScrollView (at ScrollView.js:977)
    in ScrollView (at PagerScroll.js:154)
    in PagerScroll (at TabView.js:60)
    in RCTView (at View.js:44)
    in RCTView (at View.js:44)
    in TabView (at createMaterialTopTabNavigator.js:248)
    in MaterialTabView (at createTabNavigator.js:197)
    in NavigationView (at createNavigator.js:61)
    in Navigator (at Tour/index.js:30)
    in RCTView (at View.js:44)
    in RCTView (at View.js:44)
    in RCTView (at View.js:44)
    in ImageBackground (at BackgroundTheme.js:8)
    in BackgroundTheme (at Page.js:12)
    in Page (at Tour/index.js:28)
    in TourNavigation (at SceneView.js:9)
    in SceneView (at StackViewLayout.tsx:899)
    in RCTView (at View.js:44)
    in AnimatedComponent (at StackViewCard.tsx:93)
    in RCTView (at View.js:44)
    in AnimatedComponent (at screens.native.js:59)
    in Screen (at StackViewCard.tsx:80)
    in Card (at createPointerEventsContainer.tsx:95)
    in Container (at StackViewLayout.tsx:971)
- node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:619:8 in warningWithoutStack
- node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:13205:6 in warnAboutUpdateOnUnmounted
- node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14683:33 in scheduleWork
- node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:7737:17 in enqueueForceUpdate
*- node_modules/reactn/build/methods.js:24:36 in ReactNGlobalCallback
* [native code]:null in <unknown>
- node_modules/reactn/build/global-state-manager.js:162:32 in flush
- node_modules/reactn/build/global-state-manager.js:323:37 in setObject
- node_modules/reactn/build/set-global.js:6:38 in setGlobal*
* Pages/Tours/Tour/Tickets.js:62:25 in componentDidMount$
- node_modules/@babel/runtime/node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/@babel/runtime/node_modules/regenerator-runtime/runtime.js:271:30 in invoke
- node_modules/@babel/runtime/node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/@babel/runtime/node_modules/regenerator-runtime/runtime.js:135:28 in invoke
- node_modules/@babel/runtime/node_modules/regenerator-runtime/runtime.js:145:19 in <unknown>
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in <unknown>
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:152:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:200:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:464:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:320:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:135:6 in <unknown>
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:297:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:134:17 in flushedQueue
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue`


Funny part: the warning complains about componentwillunmount but the error from the stack is about a this.setGlobal() in the componentwillmount.

To troubleshoot I added a console log in the Mount / Unmount and before each setState or setGlobal ad this is what happens:
mount
checkBeforeSet:  true
checkBeforeSet:  true
unmount
-> Now I select a page
mount
checkBeforeSet:  true
checkBeforeSet:  true
-> Warnings

Seems that I am getting a lot of this warning whenever I have a componentwillmount Async function that eventually calls setGlobal

I hope it can help in the troubleshooting, I tested out reactn-devtools and can't wait to start using it!
Keep up with the good work and thank you very much for your efforts!
Umberto

**Comment by CharlesStover** [comment]
`setGlobal` inside `componentWillUnmount`? Interesting!

So the Component is unmounting, but `this.setGlobal` is attempting to trigger a rerender.

It seems like a race condition. SetGlobal tries to rerender before removing the subscription on unmount.

Thanks for reporting this! I'll look deeper into it.

**Comment by umbertoghio** [comment]
Hi,
almost, the setGlobal is in he componentwill/did/Mount, I don't even had a componentwillUNmount, the error is misleading....note that in the console.log that I placed I get the error after the page is loaded (and still mounted), I can try to dig deeper if you want, or try to reproduce in a simple project, it's almost as the setGlobal will trigger a force update when it is not necessary

**Comment by CharlesStover** [comment]
If you have a minimal reproduction, it would be beneficial.

**Comment by umbertoghio** [comment]
I gave it a shot with a blank project but "unfortunately" :-) everything works fine, I have a complex navigation in the main one that could trigger the issue, I'll try a reverse approach, find out what I am doing in the componentwillmount that causes the issue, if I find more clues I'll happily share them

**Comment by stahlmanDesign** [comment]
I see this problem too when I change tabs, which unmounts some components. I will try to create a minimal reproduction.

**Comment by umbertoghio** [comment]
I can confirm the change is between version 1.0.0 and 2.0.0, seems to happen when a component is mounted and subscribed to the global state, then unmounted and something else changes the global state

**Comment by umbertoghio** [comment]
I tried to follow the problem in the debug, I hope it can help
![Screen Shot 2019-05-31 at 10 02 18](https://user-images.githubusercontent.com/25666241/58691344-f116b200-838b-11e9-808a-b1e3ec837888.png)

**Comment by CharlesStover** [comment]
I might need to see the actual implementation of `TimelineTrip` and its relation to `handleSelection` function.

I do notice `componentWillMount`, [which is deprecated](https://reactjs.org/docs/react-component.html#unsafe_componentwillmount). It may be both subscribing to `selectedTour` and attempting to re-render due to `selectedTour` changing before mounting. Is it possible that you can refactor `TimelineTrip` to not use `componentWillMount`?

From the above link:

>**Avoid introducing any** side-effects or **subscriptions in this method.** For those use cases, use componentDidMount() instead.

**Comment by umbertoghio** [comment]
Hi Charles, 
first of all thank you for your help!

I am still using React 16.5.0 (Latest available with Expo), but I proceeded anyway to replace all the ComponentWillMount to ComponentDidMount (adapting logic where needed).

This unfortunately haven't changed the occurrence of the problems, (still getting all the same warnings doing the same actions), but at least I won't be using deprecated code 👍 

Nice point about this.global vs getGlobal, there are indeed many places in my app where I can avoid subscription as I only need a snapshot, but in several others I do it indeed, and I found an easier error-case (the one in the mega-screenshot before was a page with multiple tabs with custom navigation  inside a component page inside another navigation).

Details:

I have a bottom bar component, used by the "Page" component (Page is just a view with a top bar and bottom bar), I need to subscribe to notification changes so I can update the badge number by tapping into this.global.notificationBadge.

The entire class is here, no async methods, no lifecycle methods, just a render:
![Screen Shot 2019-06-01 at 08 38 00](https://user-images.githubusercontent.com/25666241/58744932-a52a4280-8449-11e9-8116-dcae8658b8bf.png)

The issue occurs when from the Notifications page I tap on a notification:
this will mark the notification as read and navigate to the appropriate page.
In this case I tap on an offer (read state won't change, is a red notification) and it will navigate to that specific OfferDetail page.

Here is the exact line where it breaks in the stack trace, note that the first setGlobal worked fine, and the second don't have any callbacks.
The interesting part is that the App haven't navigated yet to the other page, it is still working in the Notification one.

![Screen Shot 2019-06-01 at 08 56 19](https://user-images.githubusercontent.com/25666241/58745036-4a91e600-844b-11e9-9d02-3cbf73bda02d.png)


Note the error about BarBottom all the way up to OfferDetails which is in the navigation tree
`Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in BarBottom (at Page.js:27)
    in RCTView (at View.js:44)
    in RCTView (at View.js:44)
    in ImageBackground (at BackgroundTheme.js:8)
    in BackgroundTheme (at Page.js:12)
    in Page (at OfferDetails.js:29)
    in OfferDetails (at SceneView.js:9)
...`

BarBottom has no lifecycle methods unless there are some hidden in Reactn
Thanks!
